### PR TITLE
WIP: Standalone repo for CSI API

### PR DESCRIPTION
**Pull request by jsafrane**
**What this PR does / why we need it**:
This is an experiment to create new `staging/src/k8s.io/csi-api` directory with CSI API. It shows how to pass new clientset to CSI volume plugin in kubelet and A/D controller.

@saad-ali, PTAL

### CODE REVIEW

File: staging/src/k8s.io/csi-api/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+Learn how to engage with the Kubernetes community on the [community
+page](http://kubernetes.io/community/).
+
+You can reach the maintainers of this repository at:
+
+- Slack: #sig-instrumention (on https://kubernetes.slack.com -- get an
+  invite at slack.kubernetes.io)
+- Mailing List:
+  https://groups.google.com/forum/#!forum/kubernetes-sig-instrumentation
```
Comment (tallclair) [inline staging/src/k8s.io/csi-api/README.md]: s/instrumentation/storage/

**Comment by verult** [comment]
I didn't have any issues coding the kubelet wiring to `CSINodeInfo`, but ran into a couple of issues while testing:
* Protobuf definitions are missing. I fixed it by adding a new package in `hack/update-generated-protobuf-dockerized.sh`.
* Got the following error: `CSINodeInfo update failed: the body of the request was in an unknown format - accepted media types include: application/json, application/yaml`
I'm guessing this is just because `csi-api` isn't vendored into `apiserver` yet.

**Comment by jsafrane** [comment]
> Protobuf definitions are missing. I fixed it by adding a new package in hack/update-generated-protobuf-dockerized.sh.

CRDs don't use protobuf, only JSON. What component wanted protobuf?

> CSINodeInfo update failed: the body of the request was in an unknown format - accepted media types include: application/json, application/yaml

CRDs don't use protobuf, only JSON.

> I'm guessing this is just because csi-api isn't vendored into apiserver yet.

It won't be ever vendored there, it's CRD.

**Comment by jsafrane** [comment]
> What component wanted protobuf?

Now I see, CRDs can be used only via application/json, but the default restclient.Config uses protobuf. I fixed the PR, see:

https://github.com/kubernetes/kubernetes/pull/67716/files#diff-e0160f895346669c2e4495acfd858ea7R612

https://github.com/kubernetes/kubernetes/pull/67716/files#diff-0ca8b7870d367d5fc309a2e737a6eed2R196

**Comment by saad-ali** [comment]
Replaced with https://github.com/kubernetes/kubernetes/pull/67803

**Comment by jsafrane** [comment]
/close
This is now part of #67803

**Pull request closed without merging**